// impor/out/index.js
'use strict';
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const path_1 = require("path");
function impor(id, rootDir) {
    if (/^\./.test(id)) {
        id = path_1.join(rootDir, id);
    }
    if (ImportJob.importQueue.indexOf(id) === -1) {
        ImportJob.importQueue.push(id);
    }
    let mod;
    return new Proxy((() => { }), {
        apply: (_, thisArg, argumentsList) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.apply(mod, thisArg, argumentsList);
        },
        construct: (_, argumentsList) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.construct(mod, argumentsList);
        },
        defineProperty: (_, key, attributes) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.defineProperty(mod, key, attributes);
        },
        deleteProperty: (_, key) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.deleteProperty(mod, key);
        },
        get: (_, key, reciver) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.get(mod, key, reciver);
        },
        getPrototypeOf: (_) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.getPrototypeOf(mod);
        },
        has: (_, key) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.has(mod, key);
        },
        isExtensible: (_) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.isExtensible(mod);
        },
        ownKeys: (_) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return mod ? Reflect.ownKeys(mod) : [];
        },
        preventExtensions: (_) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.preventExtensions(mod);
        },
        set: (_, key, value, reciver) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.set(mod, key, value, reciver);
        },
        setPrototypeOf: (_, value) => {
            if (!mod) {
                ImportJob.removeJob(id);
                mod = require(id);
            }
            return Reflect.setPrototypeOf(mod, value);
        }
    });
}
class ImportJob {
    static importJob() {
        return __awaiter(this, void 0, void 0, function* () {
            if (ImportJob.importJobBusy) {
                return;
            }
            if (ImportJob.importQueue.length === 0) {
                return;
            }
            const id = ImportJob.importQueue.shift();
            if (!id) {
                return;
            }
            ImportJob.importJobBusy = true;
            yield Promise.resolve().then(() => require(id));
            ImportJob.importJobBusy = false;
        });
    }
    static removeJob(id) {
        const index = ImportJob.importQueue.indexOf(id);
        if (index === -1) {
            return;
        }
        ImportJob.importQueue.splice(index, 1);
    }
    static start() {
        if (ImportJob.isStarted) {
            return;
        }
        ImportJob.isStarted = true;
        setInterval(ImportJob.importJob, 1000);
    }
}
ImportJob.isStarted = false;
ImportJob.importQueue = [];
ImportJob.importJobBusy = false;
ImportJob.start();
module.exports = (rootDir) => {
    return (id) => {
        return impor(id, rootDir);
    };
};
// SIG // Begin signature block
// SIG // MIIn5AYJKoZIhvcNAQcCoIIn1TCCJ9ECAQExDzANBglg
// SIG // hkgBZQMEAgEFADB3BgorBgEEAYI3AgEEoGkwZzAyBgor
// SIG // BgEEAYI3AgEeMCQCAQEEEBDgyQbOONQRoqMAEEvTUJAC
// SIG // AQACAQACAQACAQACAQAwMTANBglghkgBZQMEAgEFAAQg
// SIG // sygLtaT7wF2/NPBw/4MdTDHKba4WjOQvLRL9afPvJZyg
// SIG // gg2aMIIGGDCCBACgAwIBAgITMwAAAvQbjHGw5R+wOgAA
// SIG // AAAC9DANBgkqhkiG9w0BAQsFADB+MQswCQYDVQQGEwJV
// SIG // UzETMBEGA1UECBMKV2FzaGluZ3RvbjEQMA4GA1UEBxMH
// SIG // UmVkbW9uZDEeMBwGA1UEChMVTWljcm9zb2Z0IENvcnBv
// SIG // cmF0aW9uMSgwJgYDVQQDEx9NaWNyb3NvZnQgQ29kZSBT
// SIG // aWduaW5nIFBDQSAyMDExMB4XDTIyMDYzMDE3MzgyM1oX
// SIG // DTIzMDkxNTE3MzgyM1owgYgxCzAJBgNVBAYTAlVTMRMw
// SIG // EQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRt
// SIG // b25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRp
// SIG // b24xMjAwBgNVBAMTKU1pY3Jvc29mdCAzcmQgUGFydHkg
// SIG // QXBwbGljYXRpb24gQ29tcG9uZW50MIIBIjANBgkqhkiG
// SIG // 9w0BAQEFAAOCAQ8AMIIBCgKCAQEAovJsp2fpXM10kbg5
// SIG // LdT94DW31Bcki5WOHMgTSANJlsLg6q4++dmjVdGJWjK0
// SIG // vlVCKplUrfHIiAjhGXXc1uLZSfuW5m7eOG1Y80NUiGC3
// SIG // LybeP4dUH1VWvfzeJzKw1qgwInTGgqUzO5aOwznRngdh
// SIG // YqCnkEGv9J+jbz301ZZUgVfsnOoxid7leJqcCXdXlmoy
// SIG // rITd4JPl92bZ9j2r9foO4v9WLSz8pqsBYNSlXSWw2BOd
// SIG // VcEs+bRkqc0wr43GLbVcQHJqnUUKEsnXoWRPWkb/uaX1
// SIG // g6lmnIpXTojPxkJDLXVoaSq3jeebr+SdTWzLjHVksEw7
// SIG // YvQvw2OBbTap1koEHQIDAQABo4IBgjCCAX4wHwYDVR0l
// SIG // BBgwFgYKKwYBBAGCN0wRAQYIKwYBBQUHAwMwHQYDVR0O
// SIG // BBYEFF8RHsb5v7Dszq4BlUon5WTGyfpvMFQGA1UdEQRN
// SIG // MEukSTBHMS0wKwYDVQQLEyRNaWNyb3NvZnQgSXJlbGFu
// SIG // ZCBPcGVyYXRpb25zIExpbWl0ZWQxFjAUBgNVBAUTDTIz
// SIG // MTUyMis0NzEzMzcwHwYDVR0jBBgwFoAUSG5k5VAF04Kq
// SIG // Fzc3IrVtqMp1ApUwVAYDVR0fBE0wSzBJoEegRYZDaHR0
// SIG // cDovL3d3dy5taWNyb3NvZnQuY29tL3BraW9wcy9jcmwv
// SIG // TWljQ29kU2lnUENBMjAxMV8yMDExLTA3LTA4LmNybDBh
// SIG // BggrBgEFBQcBAQRVMFMwUQYIKwYBBQUHMAKGRWh0dHA6
// SIG // Ly93d3cubWljcm9zb2Z0LmNvbS9wa2lvcHMvY2VydHMv
// SIG // TWljQ29kU2lnUENBMjAxMV8yMDExLTA3LTA4LmNydDAM
// SIG // BgNVHRMBAf8EAjAAMA0GCSqGSIb3DQEBCwUAA4ICAQBs
// SIG // FhfFnls1uOsWNyCPzprgpge3+F16IK5Vid3DKAAP6L0Q
// SIG // ILYV5nwSulqXjDGAAjbKXGXLR/PqpRVnXKFP5nT/SscI
// SIG // iJ/h4TQ0egeauXHR0tIRJLm1rkFdmBs5rhmjDcA9WxSa
// SIG // CG1GB5UIBTYO5Tck6dVHGxdJG6/LVBam+lkwL66GcPoV
// SIG // njQqxEQ1x5BRdtu0OvTcaooz4URjQPcYIVMe4avXDOUQ
// SIG // CcTWynmtyZvonzqLfVWimbEIwhrg30/d5Nf0xClqpWIQ
// SIG // 1dFdWmku8aZDMa/RQwb0hylU/EJ6TwcZeWdaJew41yWe
// SIG // PH/CMuUtreCSu0F8948ldiZM+lAfgTNk4P3nVw5oIdRc
// SIG // h5G2fMcNAoBmQ4tBk7X0GjFwN95epDK1LQPzN82tG3la
// SIG // qdhbf7jNH/Gx++1kDTq0MTNpc53CJOf0FcMOBcmhxJ1C
// SIG // EIyHhGdI0feDmTA7i8giO5lC2gP/6vKAZpLOyW3LTxfH
// SIG // uAda1z9UW+oAxcZc6eM21Gdk1bjoz0kQDZWhtwgipD0C
// SIG // U+JXxos6501tZtlanyREh0u16IiN79Ii5g3nJ52Tnkvu
// SIG // Qv47ImKxWzFS62RAMqaX3TFYzut9R0Gn2KGvX0OGVFPR
// SIG // kuo3YLybJNAdTz0tgEN4wH9UOuGrHYaL0walq23nDdsg
// SIG // FwBXdWGJ94cV7NMw2Kdc2DCCB3owggVioAMCAQICCmEO
// SIG // kNIAAAAAAAMwDQYJKoZIhvcNAQELBQAwgYgxCzAJBgNV
// SIG // BAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYD
// SIG // VQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNyb3NvZnQg
// SIG // Q29ycG9yYXRpb24xMjAwBgNVBAMTKU1pY3Jvc29mdCBS
// SIG // b290IENlcnRpZmljYXRlIEF1dGhvcml0eSAyMDExMB4X
// SIG // DTExMDcwODIwNTkwOVoXDTI2MDcwODIxMDkwOVowfjEL
// SIG // MAkGA1UEBhMCVVMxEzARBgNVBAgTCldhc2hpbmd0b24x
// SIG // EDAOBgNVBAcTB1JlZG1vbmQxHjAcBgNVBAoTFU1pY3Jv
// SIG // c29mdCBDb3Jwb3JhdGlvbjEoMCYGA1UEAxMfTWljcm9z
// SIG // b2Z0IENvZGUgU2lnbmluZyBQQ0EgMjAxMTCCAiIwDQYJ
// SIG // KoZIhvcNAQEBBQADggIPADCCAgoCggIBAKvw+nIQHC6t
// SIG // 2G6qghBNNLrytlghn0IbKmvpWlCquAY4GgRJun/DDB7d
// SIG // N2vGEtgL8DjCmQawyDnVARQxQtOJDXlkh36UYCRsr55J
// SIG // nOloXtLfm1OyCizDr9mpK656Ca/XllnKYBoF6WZ26DJS
// SIG // JhIv56sIUM+zRLdd2MQuA3WraPPLbfM6XKEW9Ea64Dhk
// SIG // rG5kNXimoGMPLdNAk/jj3gcN1Vx5pUkp5w2+oBN3vpQ9
// SIG // 7/vjK1oQH01WKKJ6cuASOrdJXtjt7UORg9l7snuGG9k+
// SIG // sYxd6IlPhBryoS9Z5JA7La4zWMW3Pv4y07MDPbGyr5I4
// SIG // ftKdgCz1TlaRITUlwzluZH9TupwPrRkjhMv0ugOGjfdf
// SIG // 8NBSv4yUh7zAIXQlXxgotswnKDglmDlKNs98sZKuHCOn
// SIG // qWbsYR9q4ShJnV+I4iVd0yFLPlLEtVc/JAPw0XpbL9Uj
// SIG // 43BdD1FGd7P4AOG8rAKCX9vAFbO9G9RVS+c5oQ/pI0m8
// SIG // GLhEfEXkwcNyeuBy5yTfv0aZxe/CHFfbg43sTUkwp6uO
// SIG // 3+xbn6/83bBm4sGXgXvt1u1L50kppxMopqd9Z4DmimJ4
// SIG // X7IvhNdXnFy/dygo8e1twyiPLI9AN0/B4YVEicQJTMXU
// SIG // pUMvdJX3bvh4IFgsE11glZo+TzOE2rCIF96eTvSWsLxG
// SIG // oGyY0uDWiIwLAgMBAAGjggHtMIIB6TAQBgkrBgEEAYI3
// SIG // FQEEAwIBADAdBgNVHQ4EFgQUSG5k5VAF04KqFzc3IrVt
// SIG // qMp1ApUwGQYJKwYBBAGCNxQCBAweCgBTAHUAYgBDAEEw
// SIG // CwYDVR0PBAQDAgGGMA8GA1UdEwEB/wQFMAMBAf8wHwYD
// SIG // VR0jBBgwFoAUci06AjGQQ7kUBU7h6qfHMdEjiTQwWgYD
// SIG // VR0fBFMwUTBPoE2gS4ZJaHR0cDovL2NybC5taWNyb3Nv
// SIG // ZnQuY29tL3BraS9jcmwvcHJvZHVjdHMvTWljUm9vQ2Vy
// SIG // QXV0MjAxMV8yMDExXzAzXzIyLmNybDBeBggrBgEFBQcB
// SIG // AQRSMFAwTgYIKwYBBQUHMAKGQmh0dHA6Ly93d3cubWlj
// SIG // cm9zb2Z0LmNvbS9wa2kvY2VydHMvTWljUm9vQ2VyQXV0
// SIG // MjAxMV8yMDExXzAzXzIyLmNydDCBnwYDVR0gBIGXMIGU
// SIG // MIGRBgkrBgEEAYI3LgMwgYMwPwYIKwYBBQUHAgEWM2h0
// SIG // dHA6Ly93d3cubWljcm9zb2Z0LmNvbS9wa2lvcHMvZG9j
// SIG // cy9wcmltYXJ5Y3BzLmh0bTBABggrBgEFBQcCAjA0HjIg
// SIG // HQBMAGUAZwBhAGwAXwBwAG8AbABpAGMAeQBfAHMAdABh
// SIG // AHQAZQBtAGUAbgB0AC4gHTANBgkqhkiG9w0BAQsFAAOC
// SIG // AgEAZ/KGpZjgVHkaLtPYdGcimwuWEeFjkplCln3SeQyQ
// SIG // wWVfLiw++MNy0W2D/r4/6ArKO79HqaPzadtjvyI1pZdd
// SIG // ZYSQfYtGUFXYDJJ80hpLHPM8QotS0LD9a+M+By4pm+Y9
// SIG // G6XUtR13lDni6WTJRD14eiPzE32mkHSDjfTLJgJGKsKK
// SIG // ELukqQUMm+1o+mgulaAqPyprWEljHwlpblqYluSD9MCP
// SIG // 80Yr3vw70L01724lruWvJ+3Q3fMOr5kol5hNDj0L8giJ
// SIG // 1h/DMhji8MUtzluetEk5CsYKwsatruWy2dsViFFFWDgy
// SIG // cScaf7H0J/jeLDogaZiyWYlobm+nt3TDQAUGpgEqKD6C
// SIG // PxNNZgvAs0314Y9/HG8VfUWnduVAKmWjw11SYobDHWM2
// SIG // l4bf2vP48hahmifhzaWX0O5dY0HjWwechz4GdwbRBrF1
// SIG // HxS+YWG18NzGGwS+30HHDiju3mUv7Jf2oVyW2ADWoUa9
// SIG // WfOXpQlLSBCZgB/QACnFsZulP0V3HjXG0qKin3p6IvpI
// SIG // lR+r+0cjgPWe+L9rt0uX4ut1eBrs6jeZeRhL/9azI2h1
// SIG // 5q/6/IvrC4DqaTuv/DDtBEyO3991bWORPdGdVk5Pv4BX
// SIG // IqF4ETIheu9BCrE/+6jMpF3BoYibV3FWTkhFwELJm3Zb
// SIG // CoBIa/15n8G9bW1qyVJzEw16UM0xghmiMIIZngIBATCB
// SIG // lTB+MQswCQYDVQQGEwJVUzETMBEGA1UECBMKV2FzaGlu
// SIG // Z3RvbjEQMA4GA1UEBxMHUmVkbW9uZDEeMBwGA1UEChMV
// SIG // TWljcm9zb2Z0IENvcnBvcmF0aW9uMSgwJgYDVQQDEx9N
// SIG // aWNyb3NvZnQgQ29kZSBTaWduaW5nIFBDQSAyMDExAhMz
// SIG // AAAC9BuMcbDlH7A6AAAAAAL0MA0GCWCGSAFlAwQCAQUA
// SIG // oIGuMBkGCSqGSIb3DQEJAzEMBgorBgEEAYI3AgEEMBwG
// SIG // CisGAQQBgjcCAQsxDjAMBgorBgEEAYI3AgEVMC8GCSqG
// SIG // SIb3DQEJBDEiBCBKDaH88hb/xWHWKVwN72Vl21rDo9BA
// SIG // QSNf0uaJQTdeATBCBgorBgEEAYI3AgEMMTQwMqAUgBIA
// SIG // TQBpAGMAcgBvAHMAbwBmAHShGoAYaHR0cDovL3d3dy5t
// SIG // aWNyb3NvZnQuY29tMA0GCSqGSIb3DQEBAQUABIIBACQc
// SIG // vnMMLfHWShOezmBPOH+DmZgDW30AHWHcujjJrYKADx/N
// SIG // RBtMbu9QBDtIZ7Aqge2KGq5qF50uvO4t4Efrz4gVoQ28
// SIG // 42eqHYIt11PHOzOZ90+X8mDAY7vVWFsgYP+GsvJ0lBh9
// SIG // sJi5SopZrpzD4Y6yyUvsxpBb5SIzoHmqbBhM6HUubTtn
// SIG // cAUzpAiNLi2I+UpqBxj/LUmnZbmTfVGiC0CbTm0rYmmU
// SIG // SwlBtRuMhrOM5WuMx3f4XphMy7wK+cXzUnshVqr8EKV/
// SIG // 6pztoZ0+NXJaYp+RTS01kux0D8dxV0gsgotrH1wxUqy1
// SIG // 7KZ5hUihtl84qToSpPW36RVsgz07E7OhghcsMIIXKAYK
// SIG // KwYBBAGCNwMDATGCFxgwghcUBgkqhkiG9w0BBwKgghcF
// SIG // MIIXAQIBAzEPMA0GCWCGSAFlAwQCAQUAMIIBWQYLKoZI
// SIG // hvcNAQkQAQSgggFIBIIBRDCCAUACAQEGCisGAQQBhFkK
// SIG // AwEwMTANBglghkgBZQMEAgEFAAQgm+S4YY/yTCOc5hUs
// SIG // yjzCNmXDrFS1lo5jMPnl6i5YbwsCBmP3WuqvpRgTMjAy
// SIG // MzAzMTUyMTA2NDEuOTY0WjAEgAIB9KCB2KSB1TCB0jEL
// SIG // MAkGA1UEBhMCVVMxEzARBgNVBAgTCldhc2hpbmd0b24x
// SIG // EDAOBgNVBAcTB1JlZG1vbmQxHjAcBgNVBAoTFU1pY3Jv
// SIG // c29mdCBDb3Jwb3JhdGlvbjEtMCsGA1UECxMkTWljcm9z
// SIG // b2Z0IElyZWxhbmQgT3BlcmF0aW9ucyBMaW1pdGVkMSYw
// SIG // JAYDVQQLEx1UaGFsZXMgVFNTIEVTTjpBMjQwLTRCODIt
// SIG // MTMwRTElMCMGA1UEAxMcTWljcm9zb2Z0IFRpbWUtU3Rh
// SIG // bXAgU2VydmljZaCCEXswggcnMIIFD6ADAgECAhMzAAAB
// SIG // uAjUwbh54FFJAAEAAAG4MA0GCSqGSIb3DQEBCwUAMHwx
// SIG // CzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9u
// SIG // MRAwDgYDVQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNy
// SIG // b3NvZnQgQ29ycG9yYXRpb24xJjAkBgNVBAMTHU1pY3Jv
// SIG // c29mdCBUaW1lLVN0YW1wIFBDQSAyMDEwMB4XDTIyMDky
// SIG // MDIwMjIxNloXDTIzMTIxNDIwMjIxNlowgdIxCzAJBgNV
// SIG // BAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYD
// SIG // VQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNyb3NvZnQg
// SIG // Q29ycG9yYXRpb24xLTArBgNVBAsTJE1pY3Jvc29mdCBJ
// SIG // cmVsYW5kIE9wZXJhdGlvbnMgTGltaXRlZDEmMCQGA1UE
// SIG // CxMdVGhhbGVzIFRTUyBFU046QTI0MC00QjgyLTEzMEUx
// SIG // JTAjBgNVBAMTHE1pY3Jvc29mdCBUaW1lLVN0YW1wIFNl
// SIG // cnZpY2UwggIiMA0GCSqGSIb3DQEBAQUAA4ICDwAwggIK
// SIG // AoICAQCcG7H8ERxEZ+QveUDxp9+7SebzomPWlo3iA8aY
// SIG // Gl2mBoqrMNBtHLwit2Lpa6nphH9gpBknXDiQBfvhKoTk
// SIG // qqztd5xnI6sOoRTZ7BsCgDcs3XcJzjtI89ZXwSCLu+02
// SIG // 3zb52lm8EkbtA0xhwO5+z0dDGCfWwvx6KUraWxX+tOiy
// SIG // nE1p8GlLhHQIcehQKoNa5ILoIrrmmsSssUZHYxWSyhVx
// SIG // Fy/adNT7cEtHxur2YkVZxCoHmpdk3i73QCKmqQcJbT2I
// SIG // n4Kp6e2eFSTHD6NE1pkTVRTnB0TcFh3oTfcYOroCBlz6
// SIG // O4TKVqz4IOJqBztnqXNE9+q4JvYZQ65dvGxuU2pJlezS
// SIG // ffJGxeZYLZa6ME+nnMBFbbi3eGtIM2KoWp1u82hWZw37
// SIG // ecjHscufYnxE3c6cAK49pyQMiLIqtT2N1z5tplDC5QAp
// SIG // LKnpRyn1ooSvxXocP5s4qLwVcJJK/xtBdfTloEfjAVCo
// SIG // n8KTskIiwMa6r1Z3wmOEIyQmYhyLkteRwkP0yLf0I+Md
// SIG // Vr0WtVKo55qS7uP8gWFmbHb2aKvElyO4jrO5rnnACXhJ
// SIG // GJTOojpDIoIcxIY0CvU4T6aALZbl2y+6iBxq3xZR5l/+
// SIG // LKHO449Y/Lgd9VN+ICZQdl6YYzGk/hPjvUiJpP+hkVfU
// SIG // pvAvPBZ5ptlXcDJbgjQFMzxY37x8dwIDAQABo4IBSTCC
// SIG // AUUwHQYDVR0OBBYEFMT9RyU3hcMzWKNZ437nF7lH0Glu
// SIG // MB8GA1UdIwQYMBaAFJ+nFV0AXmJdg/Tl0mWnG1M1Gely
// SIG // MF8GA1UdHwRYMFYwVKBSoFCGTmh0dHA6Ly93d3cubWlj
// SIG // cm9zb2Z0LmNvbS9wa2lvcHMvY3JsL01pY3Jvc29mdCUy
// SIG // MFRpbWUtU3RhbXAlMjBQQ0ElMjAyMDEwKDEpLmNybDBs
// SIG // BggrBgEFBQcBAQRgMF4wXAYIKwYBBQUHMAKGUGh0dHA6
// SIG // Ly93d3cubWljcm9zb2Z0LmNvbS9wa2lvcHMvY2VydHMv
// SIG // TWljcm9zb2Z0JTIwVGltZS1TdGFtcCUyMFBDQSUyMDIw
// SIG // MTAoMSkuY3J0MAwGA1UdEwEB/wQCMAAwFgYDVR0lAQH/
// SIG // BAwwCgYIKwYBBQUHAwgwDgYDVR0PAQH/BAQDAgeAMA0G
// SIG // CSqGSIb3DQEBCwUAA4ICAQCfbpPJosKTKCsNw/feqYhM
// SIG // 23oABsZAAR2J9rz9oW6p5EvVPe7P+kJcmToRhbFbnWoi
// SIG // 3kWWU7GhuYUdKArgSDWf5XpaOccx3PpgTqQV6cUmltYa
// SIG // qMWgi7FR9RAbc+4p9uR48vno7gXIpR+hGdGbTcZlhiEM
// SIG // /EdALktE4+FYCVxyJVz/XVToshqPVXpa5PhRsfwQvohL
// SIG // gminfyLMqRz4glAoedgxnPdbNku5XUMdR+ApYzgLWpw3
// SIG // 270nowEli6P7N8NFzAGw7q1jZ7NF4nQBdkZy9T2saDss
// SIG // /VWGpDRiuBd/7iUWZ1YG7AmLsDV9QZksDNWWz0p4IDUh
// SIG // k2cfxUNtCY/7paytK8gG7zWzW+JJGkuGu+u4nwqr1DhS
// SIG // 5VE/DzeN4YZXcSOshzzDliLSaRByGQYkzSm5QbGGyK4I
// SIG // W9WJvbX0rC2uWSQ81WTZ3UX4VKiTslxfglZvTrVZhQxv
// SIG // ZCMDOkAF/ENIn+9tc+FTs2Tbw/JLYNZSPl414FwyVbN4
// SIG // eO7DLvRl0mM5Mvu3bYJnN4kT5HWt2FUXZjybTRTdDS8m
// SIG // 3LLBO74RQoo++XgxkARatAOClRurgXa+lE1sBNFSh8Qc
// SIG // 9gaRr5+wQrPucsZxfh1eglInJA6qj4vyCO2bLHfRGz/b
// SIG // s4+JbpXVwwD61hrXRqvsCtKHZRjVbgjMbC94Z/PiojvV
// SIG // IjCCB3EwggVZoAMCAQICEzMAAAAVxedrngKbSZkAAAAA
// SIG // ABUwDQYJKoZIhvcNAQELBQAwgYgxCzAJBgNVBAYTAlVT
// SIG // MRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdS
// SIG // ZWRtb25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9y
// SIG // YXRpb24xMjAwBgNVBAMTKU1pY3Jvc29mdCBSb290IENl
// SIG // cnRpZmljYXRlIEF1dGhvcml0eSAyMDEwMB4XDTIxMDkz
// SIG // MDE4MjIyNVoXDTMwMDkzMDE4MzIyNVowfDELMAkGA1UE
// SIG // BhMCVVMxEzARBgNVBAgTCldhc2hpbmd0b24xEDAOBgNV
// SIG // BAcTB1JlZG1vbmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBD
// SIG // b3Jwb3JhdGlvbjEmMCQGA1UEAxMdTWljcm9zb2Z0IFRp
// SIG // bWUtU3RhbXAgUENBIDIwMTAwggIiMA0GCSqGSIb3DQEB
// SIG // AQUAA4ICDwAwggIKAoICAQDk4aZM57RyIQt5osvXJHm9
// SIG // DtWC0/3unAcH0qlsTnXIyjVX9gF/bErg4r25PhdgM/9c
// SIG // T8dm95VTcVrifkpa/rg2Z4VGIwy1jRPPdzLAEBjoYH1q
// SIG // UoNEt6aORmsHFPPFdvWGUNzBRMhxXFExN6AKOG6N7dcP
// SIG // 2CZTfDlhAnrEqv1yaa8dq6z2Nr41JmTamDu6GnszrYBb
// SIG // fowQHJ1S/rboYiXcag/PXfT+jlPP1uyFVk3v3byNpOOR
// SIG // j7I5LFGc6XBpDco2LXCOMcg1KL3jtIckw+DJj361VI/c
// SIG // +gVVmG1oO5pGve2krnopN6zL64NF50ZuyjLVwIYwXE8s
// SIG // 4mKyzbnijYjklqwBSru+cakXW2dg3viSkR4dPf0gz3N9
// SIG // QZpGdc3EXzTdEonW/aUgfX782Z5F37ZyL9t9X4C626p+
// SIG // Nuw2TPYrbqgSUei/BQOj0XOmTTd0lBw0gg/wEPK3Rxjt
// SIG // p+iZfD9M269ewvPV2HM9Q07BMzlMjgK8QmguEOqEUUbi
// SIG // 0b1qGFphAXPKZ6Je1yh2AuIzGHLXpyDwwvoSCtdjbwzJ
// SIG // NmSLW6CmgyFdXzB0kZSU2LlQ+QuJYfM2BjUYhEfb3BvR
// SIG // /bLUHMVr9lxSUV0S2yW6r1AFemzFER1y7435UsSFF5PA
// SIG // PBXbGjfHCBUYP3irRbb1Hode2o+eFnJpxq57t7c+auIu
// SIG // rQIDAQABo4IB3TCCAdkwEgYJKwYBBAGCNxUBBAUCAwEA
// SIG // ATAjBgkrBgEEAYI3FQIEFgQUKqdS/mTEmr6CkTxGNSnP
// SIG // EP8vBO4wHQYDVR0OBBYEFJ+nFV0AXmJdg/Tl0mWnG1M1
// SIG // GelyMFwGA1UdIARVMFMwUQYMKwYBBAGCN0yDfQEBMEEw
// SIG // PwYIKwYBBQUHAgEWM2h0dHA6Ly93d3cubWljcm9zb2Z0
// SIG // LmNvbS9wa2lvcHMvRG9jcy9SZXBvc2l0b3J5Lmh0bTAT
// SIG // BgNVHSUEDDAKBggrBgEFBQcDCDAZBgkrBgEEAYI3FAIE
// SIG // DB4KAFMAdQBiAEMAQTALBgNVHQ8EBAMCAYYwDwYDVR0T
// SIG // AQH/BAUwAwEB/zAfBgNVHSMEGDAWgBTV9lbLj+iiXGJo
// SIG // 0T2UkFvXzpoYxDBWBgNVHR8ETzBNMEugSaBHhkVodHRw
// SIG // Oi8vY3JsLm1pY3Jvc29mdC5jb20vcGtpL2NybC9wcm9k
// SIG // dWN0cy9NaWNSb29DZXJBdXRfMjAxMC0wNi0yMy5jcmww
// SIG // WgYIKwYBBQUHAQEETjBMMEoGCCsGAQUFBzAChj5odHRw
// SIG // Oi8vd3d3Lm1pY3Jvc29mdC5jb20vcGtpL2NlcnRzL01p
// SIG // Y1Jvb0NlckF1dF8yMDEwLTA2LTIzLmNydDANBgkqhkiG
// SIG // 9w0BAQsFAAOCAgEAnVV9/Cqt4SwfZwExJFvhnnJL/Klv
// SIG // 6lwUtj5OR2R4sQaTlz0xM7U518JxNj/aZGx80HU5bbsP
// SIG // MeTCj/ts0aGUGCLu6WZnOlNN3Zi6th542DYunKmCVgAD
// SIG // sAW+iehp4LoJ7nvfam++Kctu2D9IdQHZGN5tggz1bSNU
// SIG // 5HhTdSRXud2f8449xvNo32X2pFaq95W2KFUn0CS9QKC/
// SIG // GbYSEhFdPSfgQJY4rPf5KYnDvBewVIVCs/wMnosZiefw
// SIG // C2qBwoEZQhlSdYo2wh3DYXMuLGt7bj8sCXgU6ZGyqVvf
// SIG // SaN0DLzskYDSPeZKPmY7T7uG+jIa2Zb0j/aRAfbOxnT9
// SIG // 9kxybxCrdTDFNLB62FD+CljdQDzHVG2dY3RILLFORy3B
// SIG // FARxv2T5JL5zbcqOCb2zAVdJVGTZc9d/HltEAY5aGZFr
// SIG // DZ+kKNxnGSgkujhLmm77IVRrakURR6nxt67I6IleT53S
// SIG // 0Ex2tVdUCbFpAUR+fKFhbHP+CrvsQWY9af3LwUFJfn6T
// SIG // vsv4O+S3Fb+0zj6lMVGEvL8CwYKiexcdFYmNcP7ntdAo
// SIG // GokLjzbaukz5m/8K6TT4JDVnK+ANuOaMmdbhIurwJ0I9
// SIG // JZTmdHRbatGePu1+oDEzfbzL6Xu/OHBE0ZDxyKs6ijoI
// SIG // Yn/ZcGNTTY3ugm2lBRDBcQZqELQdVTNYs6FwZvKhggLX
// SIG // MIICQAIBATCCAQChgdikgdUwgdIxCzAJBgNVBAYTAlVT
// SIG // MRMwEQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdS
// SIG // ZWRtb25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9y
// SIG // YXRpb24xLTArBgNVBAsTJE1pY3Jvc29mdCBJcmVsYW5k
// SIG // IE9wZXJhdGlvbnMgTGltaXRlZDEmMCQGA1UECxMdVGhh
// SIG // bGVzIFRTUyBFU046QTI0MC00QjgyLTEzMEUxJTAjBgNV
// SIG // BAMTHE1pY3Jvc29mdCBUaW1lLVN0YW1wIFNlcnZpY2Wi
// SIG // IwoBATAHBgUrDgMCGgMVAHBrXlahcfyG0ylxy3rlwj0T
// SIG // zA+7oIGDMIGApH4wfDELMAkGA1UEBhMCVVMxEzARBgNV
// SIG // BAgTCldhc2hpbmd0b24xEDAOBgNVBAcTB1JlZG1vbmQx
// SIG // HjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlvbjEm
// SIG // MCQGA1UEAxMdTWljcm9zb2Z0IFRpbWUtU3RhbXAgUENB
// SIG // IDIwMTAwDQYJKoZIhvcNAQEFBQACBQDnvDUgMCIYDzIw
// SIG // MjMwMzE1MjAxNDI0WhgPMjAyMzAzMTYyMDE0MjRaMHcw
// SIG // PQYKKwYBBAGEWQoEATEvMC0wCgIFAOe8NSACAQAwCgIB
// SIG // AAICF3ICAf8wBwIBAAICEU0wCgIFAOe9hqACAQAwNgYK
// SIG // KwYBBAGEWQoEAjEoMCYwDAYKKwYBBAGEWQoDAqAKMAgC
// SIG // AQACAwehIKEKMAgCAQACAwGGoDANBgkqhkiG9w0BAQUF
// SIG // AAOBgQAbP/DthZsHs+XJHw2WtN/Y7zkV2DVzuZ0pSrr0
// SIG // l0INJV7fleTW7aGQHo8pbeyrn7dkV6C/+n03wi1kAumq
// SIG // TPihfhDs0tGjCQQ0eq5L+MaJ082COlSrz0sBnF0p2GSD
// SIG // M+wyNrf/S5O0dvZeo9FdhWGM1kOPLitL0tayV4k5Ds8m
// SIG // mTGCBA0wggQJAgEBMIGTMHwxCzAJBgNVBAYTAlVTMRMw
// SIG // EQYDVQQIEwpXYXNoaW5ndG9uMRAwDgYDVQQHEwdSZWRt
// SIG // b25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9yYXRp
// SIG // b24xJjAkBgNVBAMTHU1pY3Jvc29mdCBUaW1lLVN0YW1w
// SIG // IFBDQSAyMDEwAhMzAAABuAjUwbh54FFJAAEAAAG4MA0G
// SIG // CWCGSAFlAwQCAQUAoIIBSjAaBgkqhkiG9w0BCQMxDQYL
// SIG // KoZIhvcNAQkQAQQwLwYJKoZIhvcNAQkEMSIEIFIMlp3S
// SIG // rI2Ol8oDCHUm3XESqP7yyF0PvtSFdb+F+nsOMIH6Bgsq
// SIG // hkiG9w0BCRACLzGB6jCB5zCB5DCBvQQgKOvWOKBwO0OK
// SIG // UvmNTbAX66SGE3lrD3hk2potF9Dw+x8wgZgwgYCkfjB8
// SIG // MQswCQYDVQQGEwJVUzETMBEGA1UECBMKV2FzaGluZ3Rv
// SIG // bjEQMA4GA1UEBxMHUmVkbW9uZDEeMBwGA1UEChMVTWlj
// SIG // cm9zb2Z0IENvcnBvcmF0aW9uMSYwJAYDVQQDEx1NaWNy
// SIG // b3NvZnQgVGltZS1TdGFtcCBQQ0EgMjAxMAITMwAAAbgI
// SIG // 1MG4eeBRSQABAAABuDAiBCCUzVeVYEAVyMdc3JRGg+Qm
// SIG // 7chP3WeFY6cTLhX//c+l+TANBgkqhkiG9w0BAQsFAASC
// SIG // AgCbc+RcIzU7q7yINQTDhYFhhsaVbxmYkVK9LvLQ5G78
// SIG // D6iAmb8Zu3g9gk4bakpwFR7FfOdBvyYmAWdmi6tMROmV
// SIG // jzJDFXXeglOMUFllnz6nWxg1K6EoqFhx0esCohAN806j
// SIG // ZNZDYNW1lKAExA0g5eqwuAf7RD+vWW4+fu8KPZ1Au9IE
// SIG // RiFtG7sTbcIKMCKy4DbP0UladJCfCtL23ql7w3oXjKxr
// SIG // L28MxyVHQCOFXfjpSKSdzbstq1DlaWO+/MhanP23tnbM
// SIG // 4HcIOvEOTi75G/UfJdVFaDMMPdd2Xqk6oGDTYJiFWLTH
// SIG // A04bytvMVas1cU1wPn4aM1kVq9oWd9Djn+/eLHVIO9JV
// SIG // 2S4Dr0qlIMnf5MKYYZcpJgVHCLt7x42lCtnIlRPZJFEi
// SIG // gAn8SqYbjT48DfOPPRYoFCys+qCZVI03D56DijSfWKbS
// SIG // peLmSDTjpN/590QZ/2DcibiDZ4tQ3nKzGE7D50trbaZQ
// SIG // gSkBN1j11wULVmhFh18vozw5mlegaic9/IhFjS9uCOmm
// SIG // tbwKlS9DWLpVn3ai6IK4gEKSebyoZQO+B8/lVyzDpoeC
// SIG // hL87hZqo2ZRVKrwLKOwkNPo5ada+lVwwtSqzoLnugZz7
// SIG // OyPvmq4NJwOUjPXg8TQO+bsCGFvaTDCrlXWcALdtTz+M
// SIG // LpVU4qOfqvCOMZCs9qXDi0GL5Q==
// SIG // End signature block
