// @markwhen/parser/lib/lineChecks/checkComments.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkComments = void 0;
const regex_1 = require("../regex");
const Types_1 = require("../Types");
function checkComments(line, i, lengthAtIndex, context) {
    if (line.match(regex_1.COMMENT_REGEX)) {
        const from = lengthAtIndex[i];
        const to = from + line.length;
        context.ranges.push({
            type: Types_1.RangeType.Comment,
            from,
            to,
            lineFrom: {
                line: i,
                index: 0,
            },
            lineTo: {
                line: i,
                index: line.length,
            },
        });
        const currentFoldableComment = context.currentFoldableComment();
        if (currentFoldableComment) {
            currentFoldableComment.endIndex = to;
        }
        else {
            const indexOfSlashes = line.indexOf("//");
            context.startFoldable({
                startIndex: from,
                startLine: i,
                endIndex: to,
                type: "comment",
                foldStartIndex: from + (indexOfSlashes > -1 ? indexOfSlashes + 2 : 0),
            });
        }
        return true;
    }
    else {
        context.finishFoldableComment(i);
        return false;
    }
}
exports.checkComments = checkComments;
