// @markwhen/parser/lib/lineChecks/checkTags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkTags = void 0;
const ColorUtils_1 = require("../ColorUtils");
const regex_1 = require("../regex");
const Types_1 = require("../Types");
function checkTags(line, i, lengthAtIndex, context) {
    const matches = line.matchAll(regex_1.TAG_REGEX);
    if (matches) {
        for (let m of matches) {
            if (!context.tags[m[1]]) {
                context.tags[m[1]] = ColorUtils_1.COLORS[context.paletteIndex++ % ColorUtils_1.COLORS.length];
            }
            const indexOfTag = line.indexOf("#" + m[1]);
            const from = lengthAtIndex[i] + indexOfTag;
            context.ranges.push({
                type: Types_1.RangeType.Tag,
                from,
                to: from + m[1].length + 1,
                lineFrom: {
                    line: i,
                    index: indexOfTag,
                },
                lineTo: {
                    line: i,
                    index: indexOfTag + m[1].length + 1,
                },
                content: { tag: m[1], color: context.tags[m[1]] },
            });
        }
    }
    return false;
}
exports.checkTags = checkTags;
