// @markwhen/parser/lib/lineChecks/checkTagColors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkTagColors = void 0;
const ColorUtils_1 = require("../ColorUtils");
const regex_1 = require("../regex");
const Types_1 = require("../Types");
function checkTagColors(line, i, lengthAtIndex, context) {
    const tagColorMatch = line.match(regex_1.TAG_COLOR_REGEX);
    if (tagColorMatch) {
        const tagName = tagColorMatch[1];
        const colorDef = tagColorMatch[2];
        const humanColorIndex = ColorUtils_1.HUMAN_COLORS.indexOf(colorDef);
        if (humanColorIndex === -1) {
            const rgb = (0, ColorUtils_1.hexToRgb)(colorDef);
            if (rgb) {
                context.tags[tagName] = rgb;
            }
            else {
                context.tags[tagName] = ColorUtils_1.COLORS[context.paletteIndex++ % ColorUtils_1.COLORS.length];
            }
        }
        else {
            context.tags[tagName] = ColorUtils_1.COLORS[humanColorIndex];
        }
        const indexOfTag = line.indexOf(tagName);
        const from = lengthAtIndex[i] + indexOfTag - 1;
        context.ranges.push({
            type: Types_1.RangeType.Tag,
            from,
            to: from + tagName.length + 1,
            lineFrom: {
                line: i,
                index: indexOfTag - 1,
            },
            lineTo: {
                line: i,
                index: indexOfTag + tagName.length,
            },
            content: { tag: tagName, color: context.tags[tagName] },
        });
        const indexOfColorDefPlusLength = line.indexOf(colorDef) + colorDef.length;
        context.ranges.push({
            type: Types_1.RangeType.tagDefinition,
            from,
            to: from + indexOfColorDefPlusLength,
            lineFrom: {
                line: i,
                index: indexOfTag - 1,
            },
            lineTo: {
                line: i,
                index: indexOfTag - 1 + indexOfColorDefPlusLength,
            },
            content: { tag: tagName, color: context.tags[tagName] },
        });
        return true;
    }
    return false;
}
exports.checkTagColors = checkTagColors;
