// @markwhen/parser/lib/lineChecks/checkViewers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkViewers = void 0;
const regex_1 = require("../regex");
const Types_1 = require("../Types");
function checkViewers(line, i, lengthAtIndex, context) {
    const viewersMatch = line.match(regex_1.VIEWERS_REGEX);
    if (viewersMatch) {
        const viewTagIndex = line.indexOf(viewersMatch[1]);
        context.viewers = viewersMatch[2]
            .trim()
            .split(/ |,/)
            .filter((email) => !!email && email.includes("@"));
        context.ranges.push({
            type: Types_1.RangeType.View,
            from: lengthAtIndex[i] + viewTagIndex,
            to: lengthAtIndex[i] + viewTagIndex + viewersMatch[1].length,
            lineFrom: {
                line: i,
                index: viewTagIndex,
            },
            lineTo: {
                line: i,
                index: viewTagIndex + viewersMatch[1].length,
            },
        });
        const viewerRanges = [];
        for (let j = 0; j < context.viewers.length; j++) {
            const index = line.indexOf(context.viewers[j], viewerRanges.length
                ? viewerRanges[viewerRanges.length - 1].from - lengthAtIndex[i]
                : 0);
            viewerRanges.push({
                type: Types_1.RangeType.Viewer,
                from: lengthAtIndex[i] + index,
                to: lengthAtIndex[i] + index + context.viewers[j].length,
                lineFrom: {
                    line: i,
                    index,
                },
                lineTo: {
                    line: i,
                    index: index + context.viewers[j].length,
                },
            });
        }
        context.ranges.push(...viewerRanges);
        return true;
    }
    return false;
}
exports.checkViewers = checkViewers;
