// @markwhen/parser/lib/lineChecks/checkGroupEnd.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkGroupEnd = void 0;
const regex_1 = require("../regex");
const Types_1 = require("../Types");
function checkGroupEnd(line, i, lengthAtIndex, context) {
    if (context.currentPath.length > 1 && line.match(regex_1.GROUP_END_REGEX)) {
        // We are ending our subgroup
        // context.events.push(context.eventSubgroup);
        // context.eventSubgroup = undefined;
        context.ranges.push({
            from: lengthAtIndex[i],
            to: lengthAtIndex[i] + line.length,
            type: Types_1.RangeType.Section,
            lineFrom: {
                line: i,
                index: 0,
            },
            lineTo: {
                line: i,
                index: line.length,
            },
        });
        context.endCurrentGroup(lengthAtIndex[i] + line.length, {
            line: i,
            index: line.length,
        });
        return true;
    }
    return false;
}
exports.checkGroupEnd = checkGroupEnd;
