// @markwhen/parser/lib/lineChecks/checkDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkDescription = void 0;
const regex_1 = require("../regex");
const Types_1 = require("../Types");
function checkDescription(line, i, lengthAtIndex, context) {
    const descriptionMatch = line.match(regex_1.DESCRIPTION_REGEX);
    if (descriptionMatch) {
        context.description = descriptionMatch[2];
        const descriptionTagIndex = line.indexOf(descriptionMatch[1]);
        context.ranges.push({
            type: Types_1.RangeType.Description,
            from: lengthAtIndex[i] + descriptionTagIndex,
            to: lengthAtIndex[i] + descriptionTagIndex + descriptionMatch[1].length,
            lineFrom: {
                line: i,
                index: descriptionTagIndex,
            },
            lineTo: {
                line: i,
                index: descriptionTagIndex + descriptionMatch[1].length,
            },
        });
        return true;
    }
    return false;
}
exports.checkDescription = checkDescription;
