// @markwhen/parser/lib/lineChecks/checkGroupStart.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkGroupStart = void 0;
const utils_1 = require("../dateRange/utils");
const regex_1 = require("../regex");
const Types_1 = require("../Types");
function checkGroupStart(line, i, lengthAtIndex, context) {
    const groupStart = line.match(regex_1.GROUP_START_REGEX);
    if (groupStart) {
        // We're starting a new group here.
        // Use the path to determine where it goes.
        const range = {
            from: lengthAtIndex[i],
            to: lengthAtIndex[i] + groupStart[0].length,
            type: Types_1.RangeType.Section,
            lineFrom: {
                line: i,
                index: 0,
            },
            lineTo: {
                line: i,
                index: groupStart[0].length,
            },
        };
        context.ranges.push(range);
        // This should create a new Array<Node> that we can push
        const newGroup = (0, utils_1.parseGroupFromStartTag)(line, groupStart, range);
        context.push(newGroup);
        context.startFoldableSection({
            type: Types_1.RangeType.Section,
            startLine: i,
            startIndex: lengthAtIndex[i],
            endIndex: lengthAtIndex[i] + line.length,
            foldStartIndex: lengthAtIndex[i] + line.length,
        });
        return true;
    }
    return false;
}
exports.checkGroupStart = checkGroupStart;
