// @markwhen/parser/lib/lineChecks/checkEditors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkEditors = void 0;
const regex_1 = require("../regex");
const Types_1 = require("../Types");
function checkEditors(line, i, lengthAtIndex, context) {
    const editorsMatch = line.match(regex_1.EDITORS_REGEX);
    if (!editorsMatch) {
        return false;
    }
    const editTagIndex = line.indexOf(editorsMatch[1]);
    context.editors = editorsMatch[2]
        .trim()
        .split(/ |,/)
        .filter((email) => !!email && email.includes("@"));
    context.ranges.push({
        type: Types_1.RangeType.Edit,
        from: lengthAtIndex[i] + editTagIndex,
        to: lengthAtIndex[i] + editTagIndex + editorsMatch[1].length,
        lineFrom: {
            line: i,
            index: editTagIndex,
        },
        lineTo: {
            line: i,
            index: editTagIndex + editorsMatch[1].length,
        },
    });
    const editorRanges = [];
    for (let j = 0; j < context.editors.length; j++) {
        const index = line.indexOf(context.editors[j], editorRanges.length
            ? editorRanges[editorRanges.length - 1].from - lengthAtIndex[i]
            : 0);
        editorRanges.push({
            type: Types_1.RangeType.Editor,
            from: lengthAtIndex[i] + index,
            to: lengthAtIndex[i] + index + context.editors[j].length,
            lineFrom: {
                line: i,
                index,
            },
            lineTo: {
                line: i,
                index: index + context.editors[j].length,
            },
        });
    }
    context.ranges.push(...editorRanges);
    return true;
}
exports.checkEditors = checkEditors;
