// @markwhen/parser/lib/lineChecks/checkListItems.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkListItems = void 0;
const regex_1 = require("../regex");
const Types_1 = require("../Types");
function checkListItems(line, i, lengthAtIndex, context) {
    const checklistItemMatch = line.match(regex_1.CHECKLIST_ITEM_REGEX);
    if (checklistItemMatch) {
        const from = lengthAtIndex[i];
        const indexInLine = line.indexOf(checklistItemMatch[1]) + checklistItemMatch[1].length;
        const to = from + indexInLine;
        const indicator = {
            type: Types_1.RangeType.CheckboxItemIndicator,
            from,
            to,
            lineFrom: {
                line: i,
                index: 0,
            },
            lineTo: {
                line: i,
                index: indexInLine,
            },
            content: checklistItemMatch.includes("x") || checklistItemMatch.includes("X"),
        };
        const contents = {
            type: Types_1.RangeType.ListItemContents,
            from: to,
            to: from + line.length - 1,
            lineFrom: {
                line: i,
                index: indexInLine + 1,
            },
            lineTo: {
                line: i,
                index: line.length,
            },
        };
        context.ranges.push(...[indicator, contents]);
        return [indicator, contents];
    }
    else if (line.match(regex_1.LIST_ITEM_REGEX)) {
        const from = lengthAtIndex[i];
        const indicator = {
            type: Types_1.RangeType.listItemIndicator,
            from: lengthAtIndex[i],
            to: from + 1,
            lineFrom: {
                line: i,
                index: 0,
            },
            lineTo: {
                line: i,
                index: 1,
            },
        };
        const contents = {
            type: Types_1.RangeType.ListItemContents,
            from: from,
            to: from + line.length - 1,
            lineFrom: {
                line: i,
                index: 1,
            },
            lineTo: {
                line: 1,
                index: line.length,
            },
        };
        context.ranges.push(...[indicator, contents]);
        return [indicator, contents];
    }
    return false;
}
exports.checkListItems = checkListItems;
